lib/using_restricted_array.rb

```diff
@@ -1,61 +1,109 @@
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def length(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def print_array(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil
+    print "#{array[index]} "
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |index|
+    if array[index] == value_to_find
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def find_largest(array, length)
-  raise NotImplementedError
+  larget_num = array[0]
+  length.times do |index|
+    if array[index] > larget_num
+      larget_num = array[index]
+    end
+  end
+  return larget_num
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_num = array[0]
+  length.times do |index|
+    if array[index] < smallest_num
+      smallest_num = array[index]
+    end
+  end
+  return smallest_num
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n) b/c the loop is depending on the input size (n is the size of the array)
+# Space complexity: O(1) b/c we are accessing the same array without creating a new one.
 def reverse(array, length)
```

Comment:
The method works but the time complexity is O(n) because as the array increases in size the loop runs a correspondingly increasing number of times.